src/TypeInference/Pattern.mli

```diff
@@ -16,7 +16,7 @@ open Common
   top-down in a complex patterns, while the environment flows left-to-right.
   *)
 val check_type :
-  env:Env.t -> scope:T.scope -> S.pattern -> T.typ ->
+  env:Env.t -> scope:T.scope -> ?public:bool -> S.pattern -> T.typ ->
```

Comment:
I wonder if it makes sense to allow `public` flag in patterns, i.e., in `let (pub x, y) = ...` only `x` becomes public.

Reply:
Public flag in patterns potentially would solve the problem of inelegant definition of `def` datatype (pub pub problem), and scalability to `abstract` visibility modifier. Definitions like `pub let x = 42` can be considered as a syntactic sugar for `let (pub x) = 42`, while definitions of types would store more complex visibility modifiers (pub, abstract). Or even we can have visibility modifiers to constructors, so `pub data T = A | B` would be desugared to `data pub T = pub A | pub B` while `abstract data T = A | B` would be `data pub T = A | B`.

Reply:
Therefore, I opt for getting rid of `DPub` in Surface (in Raw it still makes sense) and desugar it to public flags in patterns. I think it could be done as a separate task.